src/VisualStudio/Core/Def/Implementation/Remote/ServiceHubRemoteHostClient.ConnectionManager.cs

```diff
@@ -0,0 +1,159 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Remote;
+using Microsoft.ServiceHub.Client;
+using Roslyn.Utilities;
+
+namespace Microsoft.VisualStudio.LanguageServices.Remote
+{
+    internal sealed partial class ServiceHubRemoteHostClient : RemoteHostClient
+    {
+        private partial class ConnectionManager
+        {
+            private readonly HubClient _hubClient;
+            private readonly HostGroup _hostGroup;
+            private readonly TimeSpan _timeout;
+
+            private readonly ReaderWriterLockSlim _shutdownLock;
+            private readonly ReferenceCountedDisposable<RemotableDataJsonRpc> _remotableDataRpc;
+
+            private readonly int _maxPoolConnections;
+
+            // keyed to serviceName. each connection is for specific service such as CodeAnalysisService
+            private readonly ConcurrentDictionary<string, ConcurrentQueue<JsonRpcConnection>> _pools;
+
+            // indicate whether pool should be used.
+            // it is mutable since it will set to false when this pool got shutdown
+            private volatile bool _usePool;
```

Comment:
:bulb: Rename this to either `_enableConnectionPooling` or `_disableConnectionPooling` and change it to `readonly`.

The only time this value *changes* after it is handling for the case of initially true but set to false in `Shutdown`. This is provision is unnecessary, because the `Shutdown` method also disposes of `_remotableDataRpc`. This reference-counted disposable will never return a new reference after it is disposed, so `TryCreateNewConnectionAsync` will always fail to create a connection after shutdown.

Reply:
yep. that is the one thing I said I need to change. I will make it to create new connection if it is gone.

Reply:
decided not to do temporary connection thing. feels over engineering especially since currently we don't need it.